core/src/test/java/ai/timefold/solver/core/api/solver/SolutionManagerTest.java

```diff
@@ -386,22 +387,71 @@ void recommendFit(SolutionManagerSource SolutionManagerSource) {
 
     @ParameterizedTest
     @EnumSource(SolutionManagerSource.class)
-    void recommendFitUninitializedSolution(SolutionManagerSource SolutionManagerSource) {
+    void recommendAssignmentAlreadyAssigned(SolutionManagerSource SolutionManagerSource) {
+        int valueSize = 3;
+        var solution = TestdataShadowedSolution.generateSolution(valueSize, 3);
+        var evaluatedEntity = solution.getEntityList().get(2);
+        var value = evaluatedEntity.getValue();
+        evaluatedEntity.setValue(value);
+
+        var solutionManager = SolutionManagerSource.createSolutionManager(SOLVER_FACTORY_SHADOWED);
+        assertThat(solutionManager).isNotNull();
+        var recommendationList =
+                solutionManager.recommendAssignment(solution, evaluatedEntity, TestdataShadowedEntity::getValue);
+
+        // Three values means there need to be three recommendations.
+        assertThat(recommendationList).hasSize(valueSize);
+        /*
+         * The calculator counts how many entities have the same value as another entity.
+         * Therefore the recommendation to assign value #2 needs to come first,
+         * as it means the solution is practically unchanged.
+         */
+        var firstRecommendation = recommendationList.get(0);
+        assertSoftly(softly -> {
+            softly.assertThat(firstRecommendation.proposition()).isEqualTo(value);
+            softly.assertThat(firstRecommendation.scoreAnalysisDiff()
+                    .score()).isEqualTo(SimpleScore.of(0));
+        });
+        // The other two recommendations need to come in order of the placer; so value #0, then value #1.
+        var secondRecommendation = recommendationList.get(1);
+        assertSoftly(softly -> {
+            softly.assertThat(secondRecommendation.proposition()).isEqualTo(solution.getValueList().get(0));
+            softly.assertThat(secondRecommendation.scoreAnalysisDiff()
+                    .score()).isEqualTo(SimpleScore.of(-2));
+        });
+        var thirdRecommendation = recommendationList.get(2);
+        assertSoftly(softly -> {
+            softly.assertThat(thirdRecommendation.proposition()).isEqualTo(solution.getValueList().get(1));
+            softly.assertThat(thirdRecommendation.scoreAnalysisDiff()
+                    .score()).isEqualTo(SimpleScore.of(-2));
+        });
+        // Ensure the original solution is in its original state.
+        assertSoftly(softly -> {
```

Comment:
My comment in the `AssignmentProcessor` class relates to this validation. Please verify if I'm overlooking anything.

Reply:
The original solution will be in its original state, because it is never touched. :-) Before anything happens, the solution will be cloned, and the recommendations only ever work with the clone. This test proves that.